src/lavinmq/mqtt/session.cr

```diff
@@ -113,9 +109,12 @@ module LavinMQ
           sp = env.segment_position
           no_ack = env.message.properties.delivery_mode == 0
           if no_ack
-            delete_message(sp)
-            packet = build_packet(env, nil)
-            yield packet
+            begin
+              packet = build_packet(env, nil)
+              yield packet
+            ensure
+              delete_message(sp)
```

Comment:
you want to delete it even if we know that the delivery failed? Or do we want to requeue it then? That's what we do in AMQP

Reply:
We don't want to requeue qos=0 messages, they are fire & forget 

> 4.3.1 QoS 0: At most once delivery
The message is delivered according to the capabilities of the underlying network. No response is sent by the receiver and no retry is performed by the sender. The message arrives at the receiver either once or not at all.

Reply:
if `!no_ack`, then we will `mark_unacked(sp)` which will requeue the message.

Reply:
https://github.com/cloudamqp/lavinmq/blob/d34655fd2c3ea78ae48ef484dc36e67b9914f318/src/lavinmq/amqp/queue/queue.cr#L714-L726

Reply:
i think requeuing if we know the delivery failed (exception raised) makes more sense, at least for non clean sessions. clean sessions will delete all messages if the connection is lost, right?

Reply:
ok!